Update tests to use postgres

Developing/testing with sqlite but deploying on postgres is causing issues. Migrations don't work the same way, and assumptions about PK sequences and exception types don't hold across both databases.

Since we ultimately need to rely on postgres behaviors in the deployment, we should be running automated tests with postgres to avoid surprises.

<details>
<summary>Tests that are failing under postgres as of c7d52de9ade2a14499607fd2d86e090e96871eb8</summary>

```
$ python manage.py test
Creating test database for alias 'default'...
................WARNING:django.request:Not Found: /api/dds-projects/abc123/
...WARNING:django.request:Not Found: /api/dds-resources/
.......F..........ERROR:django.request:Internal Server Error: /api/job-answer-sets/1/create-job/
Traceback (most recent call last):
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/django/core/handlers/exception.py", line 39, in inner
    response = get_response(request)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/rest_framework/viewsets.py", line 87, in view
    return self.dispatch(request, *args, **kwargs)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/rest_framework/views.py", line 474, in dispatch
    response = self.handle_exception(exc)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/rest_framework/views.py", line 434, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/rest_framework/views.py", line 471, in dispatch
    response = handler(request, *args, **kwargs)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/django/utils/decorators.py", line 185, in inner
    return func(*args, **kwargs)
  File "/Users/dcl9/Code/python/bespin-api/data/api.py", line 307, in create_job
    job = job_factory.create_job()
  File "/Users/dcl9/Code/python/bespin-api/data/jobfactory.py", line 112, in create_job
    JobDDSOutputProject.objects.create(job=job, dds_user_credentials=worker_user_credentials)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/mock/mock.py", line 1062, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/mock/mock.py", line 1118, in _mock_call
    raise effect
ValueError: oops
..................................F........WARNING:django.request:Not Found: /api/users/current-user//109
.......................................................F....F.......................................F.
======================================================================
FAIL: testUserOnlySeeAllCredsButNoTokens (data.tests_api.DDSUserCredentialTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dcl9/Code/python/bespin-api/data/tests_api.py", line 235, in testUserOnlySeeAllCredsButNoTokens
    self.assertEqual({'id': 1, 'user': 2, 'endpoint': 1}, response.data[0])
AssertionError: {'endpoint': 1, 'id': 1, 'user': 2} != OrderedDict([('id', 7), ('user', 32), ('endpoint', 6)])

======================================================================
FAIL: test_job_restart (data.tests_api.JobsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/dcl9/Code/python/bespin-api/data/tests_api.py", line 693, in test_job_restart
    mock_make_client().restart_job.assert_called_with(str(1))
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/mock/mock.py", line 937, in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/six.py", line 718, in raise_from
    raise value
AssertionError: Expected call: restart_job('1')
Actual call: restart_job(u'47')

======================================================================
FAIL: test_jobs_cant_share_job_tokens (data.tests_models.JobTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dcl9/Code/python/bespin-api/data/tests_models.py", line 321, in test_jobs_cant_share_job_tokens
    self.assertTrue(str(raised_error.exception).startswith('UNIQUE constraint failed'))
AssertionError: False is not true

======================================================================
FAIL: test_token_must_be_unique (data.tests_models.JobTokenTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dcl9/Code/python/bespin-api/data/tests_models.py", line 601, in test_token_must_be_unique
    self.assertTrue(str(raised_error.exception).startswith('UNIQUE constraint failed'))
AssertionError: False is not true

======================================================================
FAIL: test_start_job (data.test_lando.LandoJobTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/dcl9/Code/python/bespin-api/data/test_lando.py", line 65, in test_start_job
    call(self.user, '1235', '5432')
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/mock/mock.py", line 969, in assert_has_calls
    ), cause)
  File "/Users/dcl9/Code/python/bespin-api/env/lib/python2.7/site-packages/six.py", line 718, in raise_from
    raise value
AssertionError: Calls not found.
Expected: [call(<User: test_user>, '1234', '5432'),
 call(<User: test_user>, '1235', '5432')]
Actual: [call(<User: test_user>, u'1235', u'5432'),
 call(<User: test_user>, u'1234', u'5432')]

----------------------------------------------------------------------
Ran 182 tests in 40.992s

FAILED (failures=5)
Destroying test database for alias 'default'...
```

</details>

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -4,8 +4,8 @@ jobs:
     working_directory: ~/circleci-demo-python-django
     docker:
       - image: circleci/python:2.7
-        environement:
-          DATABASE_URL: postgres://root/circle_test
+        environment:
+          DATABASE_URL: postgres://root@localhost:5432/circle_test
       - image: circleci/postgres:9.5
         environment:
           POSTGRES_USER: root

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,7 +1,6 @@
 version: 2
 jobs:
   build:
-    working_directory: ~/circleci-demo-python-django
     docker:
       - image: circleci/python:2.7
         environment:

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,0 +1,32 @@
+version: 2
+jobs:
+  build:
+    working_directory: ~/circleci-demo-python-django
+    docker:
+      - image: circleci/python:2.7
+        environement:
+          DATABASE_URL: postgres://root/circle_test
+      - image: circleci/postgres:9.5
+        environment:
+          POSTGRES_USER: root
+          POSTGRES_DB: circle_test
+    steps:
+      - checkout
+      - restore_cache:
+          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
+      - run:
+          command: |
+            virtualenv venv
+            source venv/bin/activate
+            pip install -r requirements.txt
+      - save_cache:
+          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
+          paths:
+            - "venv"
+      - run:
+          command: |
+            source venv/bin/activate
+            python manage.py test
+      - store_artifacts:
+          path: test-reports/
+          destination: tr1

--- a/bespin/settings.py
+++ b/bespin/settings.py
@@ -12,6 +12,7 @@ https://docs.djangoproject.com/en/1.10/ref/settings/
 
 import os
 from bespin.settings_base import *
+import dj_database_url
 
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = '+u!*o@^1!elg34r5u$z4xgq(_yq@fqw!d^4p6y*zh_o1buifsa'
@@ -32,6 +33,10 @@ DATABASES = {
         # 'ATOMIC_REQUESTS': True,
     }
 }
+# Update database configuration from $DATABASE_URL if set
+db_from_env = dj_database_url.config()
+DATABASES['default'].update(db_from_env)
+
 
 # So that pages served by ember development server are listed as OK to access this API
 CORS_ORIGIN_WHITELIST = (

--- a/data/api.py
+++ b/data/api.py
@@ -223,7 +223,7 @@ class DDSEndpointViewSet(viewsets.ModelViewSet):
 class DDSUserCredViewSet(viewsets.ReadOnlyModelViewSet):
     permission_classes = (permissions.IsAuthenticated,)
     serializer_class = ReadOnlyDDSUserCredSerializer
-    queryset = DDSUserCredential.objects.all()
+    queryset = DDSUserCredential.objects.all().order_by('id')
 
 
 class AdminDDSUserCredentialsViewSet(viewsets.ReadOnlyModelViewSet):

--- a/data/test_lando.py
+++ b/data/test_lando.py
@@ -63,7 +63,7 @@ class LandoJobTests(TestCase):
         mock_give_download_permissions.assert_has_calls([
             call(self.user, '1234', '5432'),
             call(self.user, '1235', '5432')
-        ])
+        ], any_order=True)
 
     @patch('data.lando.LandoJob._make_client')
     @patch('data.lando.give_download_permissions')
@@ -76,4 +76,4 @@ class LandoJobTests(TestCase):
         mock_give_download_permissions.assert_has_calls([
             call(self.user, '1234', '5432'),
             call(self.user, '1235', '5432')
-        ])
+        ], any_order=True)

--- a/data/tests_api.py
+++ b/data/tests_api.py
@@ -222,18 +222,16 @@ class DDSUserCredentialTestCase(APITestCase):
         """
         other_user = self.user_login.become_other_normal_user()
         user = self.user_login.become_normal_user()
-        self.cred = DDSUserCredential.objects.create(endpoint=self.endpoint, user=user, token='secret1',
-                                                    dds_id='1')
-        self.cred = DDSUserCredential.objects.create(endpoint=self.endpoint, user=other_user, token='secret2',
-                                                     dds_id='2')
+        cred1 = DDSUserCredential.objects.create(endpoint=self.endpoint, user=user, token='secret1', dds_id='1')
+        cred2 = DDSUserCredential.objects.create(endpoint=self.endpoint, user=other_user, token='secret2', dds_id='2')
         self.assertEqual(2, len(DDSUserCredential.objects.all()))
 
         url = reverse('ddsusercredential-list')
         response = self.client.get(url, format='json')
         self.assertEqual(response.status_code, status.HTTP_200_OK)
         self.assertEqual(2, len(response.data))
-        self.assertEqual({'id': 1, 'user': 2, 'endpoint': 1}, response.data[0])
-        self.assertEqual({'id': 2, 'user': 1, 'endpoint': 1}, response.data[1])
+        self.assertEqual({'id': cred1.id, 'user': user.id, 'endpoint': self.endpoint.id}, response.data[0])
+        self.assertEqual({'id': cred2.id, 'user': other_user.id, 'endpoint': self.endpoint.id}, response.data[1])
 
     def testUserCantCreate(self):
         user = self.user_login.become_normal_user()
@@ -690,7 +688,7 @@ class JobsTestCase(APITestCase):
         response = self.client.post(url)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
         self.assertEqual(response.data['state'], Job.JOB_STATE_RESTARTING)
-        mock_make_client().restart_job.assert_called_with(str(1))
+        mock_make_client().restart_job.assert_called_with(str(job.id))
 
         # Post to /restart/ for job in CANCEL state should work
         job.state = Job.JOB_STATE_CANCEL

--- a/data/tests_models.py
+++ b/data/tests_models.py
@@ -320,7 +320,7 @@ class JobTests(TestCase):
                                       vm_settings=self.vm_settings,
                                       vm_flavor=self.vm_flavor,
                                       )
-        self.assertTrue(str(raised_error.exception).startswith('UNIQUE constraint failed'))
+        self.assertIn("unique constraint", str(raised_error.exception).lower())
 
     def test_job_activity_creation(self):
         # Create job which should start in new state
@@ -664,7 +664,7 @@ class JobTokenTests(TestCase):
         JobToken.objects.create(token='secret1')
         with self.assertRaises(IntegrityError) as raised_error:
             JobToken.objects.create(token='secret1')
-        self.assertTrue(str(raised_error.exception).startswith('UNIQUE constraint failed'))
+        self.assertIn('unique constraint', str(raised_error.exception).lower())
 
 
 class DDSUserTests(TestCase):

--- a/requirements.txt
+++ b/requirements.txt
@@ -4,6 +4,7 @@ Django==1.10.1
 django-cors-headers==1.3.1
 django-filter==1.0.1
 djangorestframework==3.4.7
+dj-database-url==0.4.2
 DukeDSClient==0.3.18
 funcsigs==1.0.2
 gcb-web-auth==0.7

